### PR TITLE
Allow resolver errors

### PR DESCRIPTION
**Pull request by P0lip**
Here is an alternative take on error handling, this time around covering exceptions thrown by resolvers.
I believe we'll want to apply the same strategy to the parsing errors as well.
I considered doing it earlier, but decided to go with "no-throw" flow, as technically our parsers do not throw exceptions and for the most of time there is some data returned we make use of.
That being said, we can change that, and simply throw an exception if diagnostics.length > 0.

Note - we don't accumulate errors here, but this will be done in the parsing PR that has `$Ref.errors`.

What do you folks think?
@marbemac @pytlesk4

I could also add that `onError` handler Tom mentioned.
Leaving the call up to you - don't have any strong preference.

### CODE REVIEW

File: lib/util/errors.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+"use strict";
+
+const { DiagnosticSeverity } = require("@stoplight/types");
```
Comment (marbemac) [inline lib/util/errors.js]: Suggest not bringing this package in. Just set `severity` to a hardcoded val equal to that enum.

File: lib/util/errors.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+const ResolverError = exports.ResolverError = class extends Error {
+  constructor (ex, source) {
```
Comment (marbemac) [inline lib/util/errors.js]: Is just `source` enough info? Or is intention to do in another PR?

File: lib/util/errors.js (hunk 1 above)
Comment (P0lip) [reply lib/util/errors.js]: Well, I guess we could only expose:
- `path` - json path pointing at some value in the document that caused an error)
- `source` -  uri of document.
- `code` / `message`

**Comment by P0lip** [comment]
@marbemac incorporated parser error handling to this PR.
I'll cover pointer errors and we should be good to go.

File: lib/parse.js [hunk 3]
```diff
@@ -17,21 +18,21 @@ module.exports = parse;
  * The promise resolves with the parsed file contents, NOT the raw (Buffer) contents.
  */
 async function parse (path, $refs, options) {
-  try {
```
Comment (marbemac) [inline lib/parse.js]: Any reason not to put the logic below (stripHash, getExtension, etc) into the try?

File: lib/parse.js [hunk 4]
```diff
@@ -41,8 +42,29 @@ async function parse (path, $refs, options) {
...
+      throw ex;
+    }
+    else if (!options.failFast) {
+      $ref.pathType = ex.plugin.name;
+      file.data = null;
+      $ref.value = null;
+
+      const errors = $ref.errors || ($ref.errors = []);
+
+      if (ex.error instanceof StoplightParserError) {
```
Comment (marbemac) [inline lib/parse.js]: Suggest `ParserError` instead of `StoplightParserError`.

File: lib/parsers/binary.js [hunk 5]
```diff
@@ -43,7 +43,7 @@ module.exports = {
    * @param {*}      file.data      - The file contents. This will be whatever data type was returned by the resolver
    * @returns {Promise<Buffer>}
    */
-  parse (file) {
+  async parse (file) {
```
Comment (marbemac) [inline lib/parsers/binary.js]: is this required?

File: lib/pointer.js [hunk 6]
```diff
@@ -264,7 +265,7 @@ function setValue (pointer, token, value) {
     }
   }
   else {
-    throw ono.syntax(`Error assigning $ref pointer "${pointer.path}". \nCannot set "${token}" of a non-object.`);
```
Comment (marbemac) [inline lib/pointer.js]: What is `ono` doing - what do we lose by ditching it? Upstream might consider this breaking if ono provides anything beyond the basics.

Comment (P0lip) [reply lib/pointer.js]: According to docs, `ono.syntax` is just an alias for `SyntaxError`.

File: lib/parse.js (hunk 4 above)
Comment (P0lip) [reply lib/parse.js]: We've got both `ParserError` and `StoplightParserError`

File: lib/parsers/binary.js (hunk 5 above)
Comment (P0lip) [reply lib/parsers/binary.js]: In theory - it shouldn't. In practice, when all parsers do not return the promise or do not call the callback, everything blows up.
At least that was my observation - I'll take a look at it separately, and fix if needed.
For the time being, I can revert this particular change.

File: lib/parse.js (hunk 3 above)
Comment (P0lip) [reply lib/parse.js]: It's an async function anyway and we throw in `catch` (and the old code also threw), so the behavior is the same.
